plugins/gateway/tyk/v3.2.2/plugin.go

```diff
@@ -0,0 +1,190 @@
+		logger.Error("Failed to get api definition")
+		return
+	}
+	ctx.SetDefinition(r, apiDefinition)
+}
+
+// Called during response phase
+func ResponseSendTelemetry(rw http.ResponseWriter, res *http.Response, req *http.Request) {
+	logger.Info("Handling telemetry")
+	telemetry, err := createTelemetry(res, req)
+	if err != nil {
+		logger.Errorf("Failed to create telemetry: %v", err)
+		return
+	}
+
+	apiClient := newAPIClient("apiclarity.apiclarity:9000")
+	params := operations.NewPostTelemetryParams().WithBody(telemetry)
+
+	_, err = apiClient.Operations.PostTelemetry(params)
+	if err != nil {
+		logger.Errorf("Failed to post telemetry: %v", err)
+		return
+	}
+	B, _ := json.Marshal(telemetry)
+	logger.Infof("Telemetry has been sent: %s", B)
+}
+
+func createTelemetry(res *http.Response, req *http.Request) (*models.Telemetry, error) {
+	apiDefinition := ctx.GetDefinition(req)
+	if apiDefinition == nil {
+		return nil, fmt.Errorf("failed to get api definition in response")
+	}
+
+	host, port := getHostAndPortFromTargetURL(apiDefinition.Proxy.TargetURL)
+
+	reqBody, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read request body: %v", err)
+	}
+
+	resBody, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response body: %v", err)
+	}
+
+	telemetry := models.Telemetry{
+		DestinationAddress:   ":" + port, // No destination ip for now
+		DestinationNamespace: "",
+		Request: &models.Request{
+			Common: &models.Common{
+				TruncatedBody: false,
+				Body:          strfmt.Base64(reqBody),
+				Headers:       createHeaders(req.Header),
+				Version:       req.Proto,
+			},
+			Host:   host,
+			Method: req.Method,
+			Path:   req.URL.String(),
+		},
+		RequestID: uuid.NewV4().String(),
+		Response: &models.Response{
+			Common: &models.Common{
+				TruncatedBody: false,
+				Body:          strfmt.Base64(resBody),
+				Headers:       createHeaders(res.Header),
+				Version:       req.Proto,
+			},
+			StatusCode: strconv.Itoa(res.StatusCode),
+		},
+		Scheme:        req.Proto,
+		SourceAddress: req.RemoteAddr,
+	}
+
+	return &telemetry, nil
+}
+
+func getHostAndPortFromTargetURL(targetURL string) (host, port string) {
+	parsedHost, err := url.Parse(targetURL)
+	if err != nil {
+		return targetURL, ""
+	}
+	host = parsedHost.Hostname()
+	port = parsedHost.Port()
+	if port == "" {
+		port = "80"
+	}
+
+	return
+}
+
+func createHeaders(headers map[string][]string) []*models.Header {
+	ret := []*models.Header{}
+
+	// TODO support multiple values for a header
+	for header, values := range headers {
+		ret = append(ret, &models.Header{
+			Key:   header,
+			Value: values[0],
+		})
+	}
+	return ret
+}
+
+func newAPIClient(host string) *client.APIClarityPluginsTelemetriesAPI {
+	cfg := client.DefaultTransportConfig()
+	transport := httptransport.New(host, "/api", cfg.Schemes)
+	apiClient := client.New(transport, strfmt.Default)
+	return apiClient
+}
```

Comment:
Also a util function

Reply:
The reason I dont want to put it as a util function is that it will need to import the api client package to the main apiclarity go.mod